File: sdk/packages/shared/src/persistence/MigrationManager.ts
/**
 * MigrationManager - TypeScript/Node.js Database Migration Manager
 *
 * Handles incremental schema changes with automatic version tracking and rollback support.
 *
 * @module persistence
 * @author White Room Development Team
 * @copyright 2024 White Room. All rights reserved.
 */

import Database from 'better-sqlite3';
import * as fs from 'fs';
import * as path from 'path';
import { DatabaseError, DatabaseErrorCode } from './DatabaseManager';

/**
 * Defines a single database migration
 */
export interface Migration {
  /** Unique version number (must be sequential) */
  version: number;

  /** Human-readable description */
  description: string;

  /** Migration logic to execute */
  migrate: (db: Database.Database) => void;

  /** Optional rollback logic */
  rollback?: (db: Database.Database) => void;
}

/**
 * Database migration manager for schema versioning and migrations.
 *
 * **Usage:**
 * ```typescript
 * const migrationManager = new MigrationManager();
 * migrationManager.migrate(database);
 * ```
 */
export class MigrationManager {
  private migrations: Migration[];

  /**
   * Initialize with all migrations
   */
  constructor() {
    this.migrations = this.registerMigrations();
  }

  /**
   * Run all pending migrations on the database
   *
   * This method:
   * 1. Checks current schema version from metadata table
   * 2. Identifies pending migrations
   * 3. Executes each migration in a transaction
   * 4. Updates schema version after successful migration
   *
   * @param db - better-sqlite3 Database instance
   * @throws DatabaseError if migration fails
   */
  public migrate(db: Database.Database): void {
    // Get current version
    const currentVersion = this.getSchemaVersion(db);
    console.log(`[MigrationManager] Current schema version: ${currentVersion}`);

    // Filter migrations that need to run
    const pendingMigrations = this.migrations.filter((m) => m.version > currentVersion);

    if (pendingMigrations.length === 0) {
      console.log('[MigrationManager] Database is up to date');
      return;
    }

    console.log(`[MigrationManager] Found ${pendingMigrations.length} pending migration(s)`);

    // Run each migration in order
    for (const migration of pendingMigrations) {
      console.log(
        `[MigrationManager] Running migration v${migration.version}: ${migration.description}`
      );

      try {
        this.runMigration(db, migration);
        console.log(`[MigrationManager] Migration v${migration.version} completed successfully`);
      } catch (error) {
        console.error(`[MigrationManager] Migration v${migration.version} failed:`, error);
        throw new DatabaseError(
          DatabaseErrorCode.MIGRATION_FAILED,
          `Migration v${migration.version}: ${migration.description}`,
          error as Error
        );
      }
    }

    const finalVersion = Math.max(...pendingMigrations.map((m) => m.version));
    console.log(`[MigrationManager] Migration complete. New schema version: ${finalVersion}`);
  }

  /**
   * Rollback to a specific schema version
   *
   * **WARNING:** This will lose data if rolling back past schema changes.
   *
   * @param db - better-sqlite3 Database instance
   * @param targetVersion - Version to rollback to
   * @throws DatabaseError if rollback fails
   */
  public rollback(db: Database.Database, targetVersion: number): void {
    const currentVersion = this.getSchemaVersion(db);

    if (currentVersion <= targetVersion) {
      console.log(`[MigrationManager] Already at version ${targetVersion} or lower`);
      return;
    }

    console.log(`[MigrationManager] Rolling back from v${currentVersion} to v${targetVersion}`);

    // Get migrations to rollback (in reverse order)
    const migrationsToRollback = this.migrations
      .filter((m) => m.version > targetVersion && m.version <= currentVersion)
      .sort((a, b) => b.version - a.version);

    if (migrationsToRollback.length === 0) {
      console.log('[MigrationManager] No migrations to rollback');
      return;
    }

    // Run each rollback
    for (const migration of migrationsToRollback) {
      if (!migration.rollback) {
        throw new DatabaseError(
          DatabaseErrorCode.MIGRATION_FAILED,
          `Migration v${migration.version} does not support rollback`
        );
      }

      console.log(
        `[MigrationManager] Rolling back migration v${migration.version}: ${migration.description}`
      );

      try {
        migration.rollback(db);
        this.updateSchemaVersion(db, migration.version - 1);
        console.log(`[MigrationManager] Rollback v${migration.version} completed`);
      } catch (error) {
        console.error(`[MigrationManager] Rollback v${migration.version} failed:`, error);
        throw new DatabaseError(
          DatabaseErrorCode.MIGRATION_FAILED,
          `Rollback v${migration.version} failed`,
          error as Error
        );
      }
    }

    console.log(`[MigrationManager] Rollback complete. New schema version: ${targetVersion}`);
  }

  /**
   * Execute a single migration in a transaction
   *
   * @param db - better-sqlite3 Database instance
   * @param migration - Migration to execute
   * @throws DatabaseError if migration fails
   */
  private runMigration(db: Database.Database, migration: Migration): void {
    // Use immediate transaction for DDL statements
    const runMigration = db.transaction(() => {
      // Execute migration
      migration.migrate(db);

      // Update schema version
      this.updateSchemaVersion(db, migration.version);
    });

    runMigration();
  }

  /**
   * Get current schema version from metadata table
   *
   * @param db - better-sqlite3 Database instance
   * @returns Current schema version (0 if not set)
   */
  private getSchemaVersion(db: Database.Database): number {
    const stmt = db.prepare('SELECT value FROM metadata WHERE key = ?');
    const result = stmt.get('schema_version') as { value: string } | undefined;
    return result ? parseInt(result.value, 10) : 0;
  }

  /**
   * Update schema version in metadata table
   *
   * @param db - better-sqlite3 Database instance
   * @param version - New schema version
   */
  private updateSchemaVersion(db: Database.Database, version: number): void {
    const stmt = db.prepare(
      'INSERT OR REPLACE INTO metadata (key, value, updated_at) VALUES (?, ?, datetime("now"))'
    );
    stmt.run('schema_version', version.toString());
  }

  /**
   * Register all migrations in order
   *
   * **IMPORTANT:** Add new migrations here with sequential version numbers
   *
   * @returns Array of all migrations
   */
  private registerMigrations(): Migration[] {
    return [
      // Migration 1: Initial schema
      {
        version: 1,
        description:
          'Initial schema with all 9 tables (songs, performances, user_preferences, roles, sections, autosaves, backups, mix_graphs, markers)',
        migrate: (db) => {
          // Execute schema.sql file
          this.executeSchemaFile(db);
        },
      },

      // Migration 2: Add instrumentId, voiceId, presetId to TrackConfig (in mix_graph_json)
      {
        version: 2,
        description:
          'Add instrumentId, voiceId, presetId fields to songs.mix_graph_json for TrackConfig objects',
        migrate: (db) => {
          // This is a data migration - update existing mix_graph_json to include new fields
          // For now, this is a no-op since the schema already includes these fields
          // Future versions will add ALTER TABLE statements if needed
          console.log(
            '[MigrationManager] Migration v2: instrumentId/voiceId/presetId already in schema'
          );
        },
      },

      // TODO: Add future migrations here
      //
      // {
      //   version: 3,
      //   description: 'Add full-text search indexes',
      //   migrate: (db) => {
      //     db.exec('CREATE VIRTUAL TABLE IF NOT EXISTS songs_fts USING fts5(name, composer, genre)');
      //   }
      // }
    ];
  }

  /**
   * Execute schema.sql file
   *
   * @param db - better-sqlite3 Database instance
   * @throws DatabaseError if schema execution fails
   */
  private executeSchemaFile(db: Database.Database): void {
    // Try multiple paths for schema.sql
    const schemaPath = this.findSchemaFile();

    if (!schemaPath) {
      throw new DatabaseError(
        DatabaseErrorCode.SCHEMA_FILE_NOT_FOUND,
        'Schema file (schema.sql) not found'
      );
    }

    console.log(`[MigrationManager] Executing schema file: ${schemaPath}`);

    // Read schema file
    const schemaSQL = fs.readFileSync(schemaPath, 'utf-8');

    // Execute schema
    db.exec(schemaSQL);

    console.log('[MigrationManager] Schema file executed successfully');
  }

  /**
   * Find schema.sql file in multiple locations
   *
   * @returns Path to schema.sql or null if not found
   */
  private findSchemaFile(): string | null {
    const possiblePaths = [
      // Try relative to this file
      path.join(__dirname, '../../../../juce_backend/shared/persistence/schema.sql'),
      // Try from package root
      path.join(__dirname, '../../../juce_backend/shared/persistence/schema.sql'),
      // Try from project root
      path.join(process.cwd(), 'juce_backend/shared/persistence/schema.sql'),
      // Try from environment variable
      process.env.PROJECT_ROOT
        ? path.join(process.env.PROJECT_ROOT, 'juce_backend/shared/persistence/schema.sql')
        : null,
    ].filter(Boolean) as string[];

    for (const schemaPath of possiblePaths) {
      if (fs.existsSync(schemaPath)) {
        return schemaPath;
      }
    }

    return null;
  }

  /**
   * Get list of all registered migrations
   *
   * @returns Array of migration descriptions
   */
  public allMigrations(): { version: number; description: string }[] {
    return this.migrations.map((m) => ({ version: m.version, description: m.description }));
  }

  /**
   * Get pending migrations (not yet applied)
   *
   * @param db - better-sqlite3 Database instance
   * @returns Array of pending migrations
   */
  public pendingMigrations(db: Database.Database): { version: number; description: string }[] {
    const currentVersion = this.getSchemaVersion(db);
    return this.migrations
      .filter((m) => m.version > currentVersion)
      .map((m) => ({ version: m.version, description: m.description }));
  }

  /**
   * Validate that all migrations can run without errors
   *
   * This is useful for testing migrations before deployment.
   *
   * @param db - better-sqlite3 Database instance
   * @returns true if all migrations are valid
   */
  public validateMigrations(db: Database.Database): boolean {
    console.log('[MigrationManager] Validating migrations...');

    for (const migration of this.migrations) {
      try {
        // Begin test transaction (immediate for DDL)
        const testMigration = db.transaction(() => {
          // Run migration
          migration.migrate(db);
        });

        testMigration(); // Will be rolled back automatically in immediate mode

        console.log(
          `✓ Migration v${migration.version}: ${migration.description}`
        );
      } catch (error) {
        console.error(`✗ Migration v${migration.version} failed:`, error);
        return false;
      }
    }

    console.log('[MigrationManager] All migrations validated successfully');
    return true;
  }
}

/**
 * Convenience function to create migration manager
 *
 * @returns MigrationManager instance
 */
export function createMigrationManager(): MigrationManager {
  return new MigrationManager();
}

File: swift_frontend/SwiftFrontendShared/Services/MigrationManager.swift
//
//  MigrationManager.swift
//  White Room
//
//  Created by White Room Development Team
//  Copyright © 2024 White Room. All rights reserved.
//

import Foundation
import GRDB

/// Database migration manager for schema versioning and migrations.
///
/// Handles incremental schema changes with automatic version tracking and rollback support.
///
/// **Usage:**
/// ```swift
/// let migrationManager = MigrationManager()
/// try migrationManager.migrate(database)
/// ```
public actor MigrationManager {

    // MARK: - Migration Definition

    /// Defines a single database migration
    public struct Migration: Sendable {
        /// Unique version number (must be sequential)
        public let version: Int

        /// Human-readable description
        public let description: String

        /// Migration logic to execute
        public let migrate: (Database) throws -> Void

        /// Optional rollback logic
        public let rollback: ((Database) throws -> Void)?

        public init(
            version: Int,
            description: String,
            migrate: @escaping (Database) throws -> Void,
            rollback: ((Database) throws -> Void)? = nil
        ) {
            self.version = version
            self.description = description
            self.migrate = migrate
            self.rollback = rollback
        }
    }

    // MARK: - Properties

    /// All registered migrations in version order
    private let migrations: [Migration]

    // MARK: - Initialization

    /// Initialize with all migrations
    public init() {
        self.migrations = Self.registerMigrations()
    }

    // MARK: - Public API

    /// Run all pending migrations on the database
    ///
    /// This method:
    /// 1. Checks current schema version from metadata table
    /// 2. Identifies pending migrations
    /// 3. Executes each migration in a transaction
    /// 4. Updates schema version after successful migration
    ///
    /// - Parameter database: GRDB Database instance
    /// - Throws: Database errors if migration fails
    public func migrate(_ database: Database) throws {
        // Get current version
        let currentVersion = try getSchemaVersion(database: database)
        print("[MigrationManager] Current schema version: \(currentVersion)")

        // Filter migrations that need to run
        let pendingMigrations = migrations.filter { $0.version > currentVersion }

        guard !pendingMigrations.isEmpty else {
            print("[MigrationManager] Database is up to date")
            return
        }

        print("[MigrationManager] Found \(pendingMigrations.count) pending migration(s)")

        // Run each migration in order
        for migration in pendingMigrations {
            print("[MigrationManager] Running migration v\(migration.version): \(migration.description)")

            do {
                try runMigration(database: database, migration: migration)
                print("[MigrationManager] Migration v\(migration.version) completed successfully")
            } catch {
                print("[MigrationManager] Migration v\(migration.version) failed: \(error)")
                throw DatabaseError.migrationFailed("Migration v\(migration.version): \(migration.description)")
            }
        }

        let finalVersion = pendingMigrations.map { $0.version }.max() ?? currentVersion
        print("[MigrationManager] Migration complete. New schema version: \(finalVersion)")
    }

    /// Rollback to a specific schema version
    ///
    /// **WARNING:** This will lose data if rolling back past schema changes.
    ///
    /// - Parameters:
    ///   - database: GRDB Database instance
    ///   - targetVersion: Version to rollback to
    /// - Throws: Database errors if rollback fails
    public func rollback(_ database: Database, to targetVersion: Int) throws {
        let currentVersion = try getSchemaVersion(database: database)

        guard currentVersion > targetVersion else {
            print("[MigrationManager] Already at version \(targetVersion) or lower")
            return
        }

        print("[MigrationManager] Rolling back from v\(currentVersion) to v\(targetVersion)")

        // Get migrations to rollback (in reverse order)
        let migrationsToRollback = migrations
            .filter { $0.version > targetVersion && $0.version <= currentVersion }
            .sorted { $0.version > $1.version }

        guard !migrationsToRollback.isEmpty else {
            print("[MigrationManager] No migrations to rollback")
            return
        }

        // Run each rollback
        for migration in migrationsToRollback {
            guard let rollback = migration.rollback else {
                throw DatabaseError.migrationFailed("Migration v\(migration.version) does not support rollback")
            }

            print("[MigrationManager] Rolling back migration v\(migration.version): \(migration.description)")

            do {
                try rollback(database)
                try updateSchemaVersion(database: database, version: migration.version - 1)
                print("[MigrationManager] Rollback v\(migration.version) completed")
            } catch {
                print("[MigrationManager] Rollback v\(migration.version) failed: \(error)")
                throw DatabaseError.migrationFailed("Rollback v\(migration.version) failed")
            }
        }

        print("[MigrationManager] Rollback complete. New schema version: \(targetVersion)")
    }

    // MARK: - Private Methods

    /// Execute a single migration in a transaction
    ///
    /// - Parameters:
    ///   - database: GRDB Database instance
    ///   - migration: Migration to execute
    /// - Throws: Database errors if migration fails
    private func runMigration(database: Database, migration: Migration) throws {
        // Begin transaction
        try database.beginTransaction()

        do {
            // Execute migration
            try migration.migrate(database)

            // Update schema version
            try updateSchemaVersion(database: database, version: migration.version)

            // Commit transaction
            try database.commit()
        } catch {
            // Rollback on error
            try database.rollback()
            throw error
        }
    }

    /// Get current schema version from metadata table
    ///
    /// - Parameter database: GRDB Database instance
    /// - Returns: Current schema version (0 if not set)
    /// - Throws: Database errors if query fails
    private func getSchemaVersion(database: Database) throws -> Int {
        guard let versionString = try String.fetchOne(
            database,
            sql: "SELECT value FROM metadata WHERE key = 'schema_version'"
        ) else {
            return 0
        }

        return Int(versionString) ?? 0
    }

    /// Update schema version in metadata table
    ///
    /// - Parameters:
    ///   - database: GRDB Database instance
    ///   - version: New schema version
    /// - Throws: Database errors if update fails
    private func updateSchemaVersion(database: Database, version: Int) throws {
        try database.execute(
            sql: """
            INSERT OR REPLACE INTO metadata (key, value, updated_at)
            VALUES ('schema_version', ?, datetime('now'))
            """,
            arguments: [String(version)]
        )
    }

    // MARK: - Migration Registry

    /// Register all migrations in order
    ///
    /// **IMPORTANT:** Add new migrations here with sequential version numbers
    ///
    /// - Returns: Array of all migrations
    private static func registerMigrations() -> [Migration] {
        return [
            // Migration 1: Initial schema
            Migration(
                version: 1,
                description: "Initial schema with all 9 tables (songs, performances, user_preferences, roles, sections, autosaves, backups, mix_graphs, markers)"
            ) { db in
                // Execute schema.sql file
                try executeSchemaFile(database: db)
            },

            // Migration 2: Add instrumentId, voiceId, presetId to TrackConfig (in mix_graph_json)
            Migration(
                version: 2,
                description: "Add instrumentId, voiceId, presetId fields to songs.mix_graph_json for TrackConfig objects"
            ) { db in
                // This is a data migration - update existing mix_graph_json to include new fields
                // For now, this is a no-op since the schema already includes these fields
                // Future versions will add ALTER TABLE statements if needed
                print("[MigrationManager] Migration v2: instrumentId/voiceId/presetId already in schema")
            },

            // TODO: Add future migrations here
            //
            // Migration(
            //     version: 3,
            //     description: "Add full-text search indexes"
            // ) { db in
            //     try db.execute(sql: "CREATE VIRTUAL TABLE IF NOT EXISTS songs_fts USING fts5(name, composer, genre)")
            // }
        ]
    }

    /// Execute schema.sql file
    ///
    /// - Parameter database: GRDB Database instance
    /// - Throws: Database errors if schema execution fails
    private static func executeSchemaFile(database: Database) throws {
        // Try multiple paths for schema.sql
        let possiblePaths = [
            Bundle.module.path(forResource: "schema", ofType: "sql"),
            Bundle.main.path(forResource: "schema", ofType: "sql"),
            // Fallback to project-relative path
            URL(fileURLWithPath: "\(ProcessInfo.processInfo.environment["PROJECT_ROOT"] ?? ".")/juce_backend/shared/persistence/schema.sql").path
        ].compactMap { $0 }

        guard let schemaPath = possiblePaths.first(where: { FileManager.default.fileExists(atPath: $0) }) else {
            throw DatabaseError.schemaFileNotFound
        }

        print("[MigrationManager] Executing schema file: \(schemaPath)")

        // Read schema file
        let schemaSQL = try String(contentsOfFile: schemaPath, encoding: .utf8)

        // Execute schema
        try database.execute(sql: schemaSQL)

        print("[MigrationManager] Schema file executed successfully")
    }
}

// MARK: - Migration Logging

extension MigrationManager {

    /// Get list of all registered migrations
    ///
    /// - Returns: Array of migration descriptions
    public func allMigrations() -> [(version: Int, description: String)] {
        return migrations.map { ($0.version, $0.description) }
    }

    /// Get pending migrations (not yet applied)
    ///
    /// - Parameter database: GRDB Database instance
    /// - Returns: Array of pending migrations
    /// - Throws: Database errors if version check fails
    public func pendingMigrations(_ database: Database) throws -> [(version: Int, description: String)] {
        let currentVersion = try getSchemaVersion(database: database)
        return migrations
            .filter { $0.version > currentVersion }
            .map { ($0.version, $0.description) }
    }
}

// MARK: - Validation

extension MigrationManager {

    /// Validate that all migrations can run without errors
    ///
    /// This is useful for testing migrations before deployment.
    ///
    /// - Parameter database: GRDB Database instance
    /// - Returns: true if all migrations are valid
    public func validateMigrations(_ database: Database) -> Bool {
        print("[MigrationManager] Validating migrations...")

        for migration in migrations {
            do {
                // Begin test transaction
                try database.beginTransaction()

                // Run migration
                try migration.migrate(database)

                // Rollback (don't actually apply)
                try database.rollback()

                print("[MigrationManager] ✓ Migration v\(migration.version): \(migration.description)")
            } catch {
                print("[MigrationManager] ✗ Migration v\(migration.version) failed: \(error)")
                return false
            }
        }

        print("[MigrationManager] All migrations validated successfully")
        return true
    }
}
